trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/IntegerTranslation.java

```diff
@@ -419,25 +419,24 @@ public CPrimitive getCTypeOfPointerComponents() {
 	}
 
 	@Override
-	public void addAssumeValueInRangeStatements(final ILocation loc, final Expression expr, final CType cType,
-			final ExpressionResultBuilder expressionResultBuilder) {
+	public Optional<Expression> getTypeConstraint(final ILocation loc, final Expression expr, final CType cType) {
 		if (!mSettings.assumeNondeterministicValuesInRange() || !cType.getUnderlyingType().isIntegerType()) {
 			// only integer types can be out of range
-			return;
+			return Optional.empty();
 		}
 		final CPrimitive cPrimitive = (CPrimitive) CEnum.replaceEnumWithInt(cType.getUnderlyingType());
 		if (mTypeSizes.isUnsigned(cPrimitive)) {
 			// only signed types can be out of range

```

Comment:
Is this also the case for quantified variables?

I remember that at one point, we had two approaches to unsigned variables: either (1) insert appropriate modulo operations in every comparison (so variables may have values out-of-range but it cannot be observed by the program), or (2) insert appropriate modulo operations whenever an unsigned value is computed.

It seems to me that while (1) may work without type constraints in quantified expressions, (2) probably would not.

Reply:
Currently, we use (1) in `IntegerTranslation`. There was some idea for an alternative translation that uses (2), but which is not implemented on `dev` (only started on some older branch).
There is also a test for unsigned quantified variables:
`//@ assert \forall unsigned x; x >= 0;` is translated to `assert forall x :: int x % 4294967296 >= 0;`, which holds trivially.